File: apps/android/feature/library/src/main/kotlin/app/storyarc/feature/library/SmbSheet.kt
package app.storyarc.feature.library

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Folder
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import app.storyarc.core.designsystem.theme.LocalStoryArcPalette
import app.storyarc.core.designsystem.tokens.StoryArcSpace
import app.storyarc.core.model.Source

/**
 * Adding a network share.
 *
 * Two steps in one sheet: what to connect to, then which folder to read. `network-share`
 * asks for both -- the connection "validated before saving" with the specific failure named,
 * and the reader able to "browse the share's directory tree and pick the folder to use as
 * the library root".
 */
@OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)
@Composable
fun SmbSheet(
    connection: SmbConnection,
    onAdd: (Source) -> Unit,
    onDismiss: () -> Unit,
) {
    val palette = LocalStoryArcPalette.current
    val step by connection.step.collectAsStateWithLifecycle()

    ModalBottomSheet(onDismissRequest = onDismiss) {
        Column(
            modifier = Modifier
                .verticalScroll(rememberScrollState())
                .padding(horizontal = StoryArcSpace.gutter)
                .padding(bottom = StoryArcSpace.xl),
            verticalArrangement = Arrangement.spacedBy(StoryArcSpace.md),
        ) {
            Text(
                text = stringResource(R.string.smb_title),
                style = MaterialTheme.typography.headlineSmall,
                color = palette.textPrimary,
            )

            when (val current = step) {
                is SmbConnection.Step.Browsing -> Chooser(connection, current, onAdd, onDismiss)
                else -> Details(connection, current)
            }
        }
    }
}

@Composable
private fun ColumnScope.Details(connection: SmbConnection, step: SmbConnection.Step) {
    val palette = LocalStoryArcPalette.current
    val host by connection.host.collectAsStateWithLifecycle()
    val share by connection.share.collectAsStateWithLifecycle()
    val username by connection.username.collectAsStateWithLifecycle()
    val password by connection.password.collectAsStateWithLifecycle()

    OutlinedTextField(
        value = host,
        onValueChange = { connection.host.value = it },
        label = { Text(stringResource(R.string.smb_host_label)) },
        singleLine = true,
        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Uri),
        modifier = Modifier.fillMaxWidth(),
    )
    Text(
        text = stringResource(R.string.smb_host_hint),
        style = MaterialTheme.typography.bodySmall,
        color = palette.textSecondary,
    )

    OutlinedTextField(
        value = share,
        onValueChange = { connection.share.value = it },
        label = { Text(stringResource(R.string.smb_share_label)) },
        singleLine = true,
        modifier = Modifier.fillMaxWidth(),
    )

    OutlinedTextField(
        value = username,
        onValueChange = { connection.username.value = it },
        label = { Text(stringResource(R.string.smb_user_label)) },
        singleLine = true,
        modifier = Modifier.fillMaxWidth(),
    )
    Text(
        text = stringResource(R.string.smb_user_hint),
        style = MaterialTheme.typography.bodySmall,
        color = palette.textSecondary,
    )

    OutlinedTextField(
        value = password,
        onValueChange = { connection.password.value = it },
        label = { Text(stringResource(R.string.smb_password_label)) },
        singleLine = true,
        visualTransformation = PasswordVisualTransformation(),
        modifier = Modifier.fillMaxWidth(),
    )

    Button(
        onClick = { connection.connect() },
        enabled = connection.isReady() && step !is SmbConnection.Step.Connecting,
        modifier = Modifier.fillMaxWidth(),
    ) {
        Text(stringResource(R.string.smb_connect))
    }

    when (step) {
        is SmbConnection.Step.Connecting ->
            CircularProgressIndicator(strokeWidth = 2.dp, modifier = Modifier.padding(2.dp))

        // `network-share` wants the specific failure, and the four the client separates are
        // four different sentences here.
        is SmbConnection.Step.Failed -> Text(
            text = step.message,
            style = MaterialTheme.typography.bodyMedium,
            color = palette.textPrimary,
        )

        else -> Unit
    }
}

@Composable
private fun ColumnScope.Chooser(
    connection: SmbConnection,
    step: SmbConnection.Step.Browsing,
    onAdd: (Source) -> Unit,
    onDismiss: () -> Unit,
) {
    val palette = LocalStoryArcPalette.current
    val folders = step.entries.filter { it.isDirectory }

    // `network-share`: the detail screen states whether the connection is encrypted. Said
    // here too, because this is the moment a reader decides whether to trust it.
    Text(
        text = stringResource(
            if (step.identity.isEncrypted) R.string.smb_encrypted else R.string.smb_not_encrypted,
            step.identity.dialect,
        ),
        style = MaterialTheme.typography.bodySmall,
        color = palette.textSecondary,
    )

    Text(
        text = step.path.ifEmpty { stringResource(R.string.smb_root) },
        style = MaterialTheme.typography.titleMedium,
        color = palette.textPrimary,
    )

    LazyColumn(modifier = Modifier.heightIn(max = 280.dp)) {
        connection.parentOf(step.path)?.let { parent ->
            item(key = "up") {
                FolderRow(name = stringResource(R.string.catalogue_back), isUp = true) {
                    connection.enter(parent)
                }
            }
        }
        items(folders, key = { it.path }) { folder ->
            FolderRow(name = folder.name, isUp = false) { connection.enter(folder.path) }
        }
    }

    Button(
        onClick = {
            connection.source()?.let(onAdd)
            onDismiss()
        },
        modifier = Modifier.fillMaxWidth(),
    ) {
        Text(stringResource(R.string.smb_use_folder))
    }
}

@Composable
private fun FolderRow(name: String, isUp: Boolean, onOpen: () -> Unit) {
    val palette = LocalStoryArcPalette.current
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(StoryArcSpace.sm),
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onOpen)
            .defaultMinSize(minHeight = 48.dp)
            .padding(vertical = StoryArcSpace.xs),
    ) {
        Icon(
            imageVector = if (isUp) Icons.AutoMirrored.Filled.ArrowBack else Icons.Filled.Folder,
            contentDescription = null,
            tint = palette.textSecondary,
        )
        Text(name, style = MaterialTheme.typography.bodyLarge, color = palette.textPrimary)
    }
}

File: apps/ios/Packages/StoryArcKit/Sources/LibraryFeature/SmbSheet.swift
public import SwiftUI

internal import DesignSystem
public import Smb
public import StoryArcCore

/// Adding a network share.
///
/// Two steps in one sheet: what to connect to, then which folder to read. `network-share`
/// asks for both — the connection "validated before saving" with the specific failure named,
/// and the reader able to "browse the share's directory tree and pick the folder to use as
/// the library root".
public struct SmbSheet: View {
    @Environment(\.theme) private var theme
    @Environment(\.dismiss) private var dismiss

    private let connection: SmbConnection
    private let onAdd: (Source) -> Void

    /// `network-share` marks discovery a SHOULD and is firm that "manual entry is always
    /// available and never gated behind discovery". So the list sits above the form and an
    /// empty one shows nothing at all — no spinner, no "searching", no reason to wait. A
    /// refused local-network permission arrives here as simply no results.
    @State private var discovery = SmbDiscovery()

    public init(connection: SmbConnection, onAdd: @escaping (Source) -> Void) {
        self.connection = connection
        self.onAdd = onAdd
    }

    public var body: some View {
        NavigationStack {
            Form {
                if case let .browsing(identity, path, entries) = connection.step {
                    chooser(identity: identity, path: path, entries: entries)
                } else {
                    details
                }
            }
            .navigationTitle(Text("smb.title", bundle: .module))
            .task { discovery.start() }
            .onDisappear { discovery.stop() }
            #if os(iOS)
            .navigationBarTitleDisplayMode(.inline)
            #endif
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button { dismiss() } label: { Text("smb.cancel", bundle: .module) }
                }
            }
        }
    }

    @ViewBuilder
    private var details: some View {
        if !discovery.hosts.isEmpty {
            Section {
                ForEach(discovery.hosts) { host in
                    Button {
                        connection.host = host.name
                    } label: {
                        Label(host.name, systemImage: "externaldrive.badge.wifi")
                    }
                }
            } header: {
                Text("smb.found", bundle: .module)
            }
        }

        Section {
            TextField(
                String(localized: "smb.host.label", bundle: .module, locale: .storyArc),
                text: Binding(get: { connection.host }, set: { connection.host = $0 })
            )
            #if os(iOS)
            .textInputAutocapitalization(.never)
            #endif
            .autocorrectionDisabled()

            TextField(
                String(localized: "smb.share.label", bundle: .module, locale: .storyArc),
                text: Binding(get: { connection.share }, set: { connection.share = $0 })
            )
            #if os(iOS)
            .textInputAutocapitalization(.never)
            #endif
            .autocorrectionDisabled()
        } footer: {
            Text("smb.host.hint", bundle: .module)
        }

        Section {
            TextField(
                String(localized: "smb.user.label", bundle: .module, locale: .storyArc),
                text: Binding(get: { connection.username }, set: { connection.username = $0 })
            )
            #if os(iOS)
            .textInputAutocapitalization(.never)
            #endif
            .autocorrectionDisabled()

            SecureField(
                String(localized: "smb.password.label", bundle: .module, locale: .storyArc),
                text: Binding(get: { connection.password }, set: { connection.password = $0 })
            )
        } footer: {
            Text("smb.user.hint", bundle: .module)
        }

        Section {
            // **The one thing on this form that does something, shaped like it.** Left
            // unstyled inside a `Form`, this drew as a list row — a white capsule the same
            // colour, height and corner as the four field groups above it, with its label in
            // the grey a placeholder wears while it is disabled, which is the state a reader
            // meets it in. The 2026-09-02 sweep called it "a fifth thing to type into", and
            // `ios-add-share-sheet.png` is a picture of five identical white capsules.
            //
            // So the row's own background goes and the button draws its own shape.
            // `.glassProminent` is how this app emphasises: `design.md` §5 keeps chrome glass
            // untinted so it picks up what is behind it, and the prominent variant is the one
            // meant to carry a tint — `.tint` on plain `.glass` tints the *material* and
            // flattens it, which `GlassIsUntintedTests` fails the build over.
            Button {
                Task { await connection.connect() }
            } label: {
                Group {
                    if case .connecting = connection.step {
                        ProgressView()
                    } else {
                        Text("smb.connect", bundle: .module)
                    }
                }
                .frame(maxWidth: .infinity)
                .padding(.vertical, StoryArcSpace.xs)
            }
            .buttonStyle(.glassProminent)
            .controlSize(.large)
            .tint(theme.accent)
            .listRowBackground(Color.clear)
            .listRowInsets(EdgeInsets())
            .disabled(!connection.canConnect)

            // `network-share` wants the specific failure, and the four the client separates
            // are four different sentences here.
            if case let .failed(message) = connection.step {
                Text(message)
                    .textRole(.footnote)
                    .foregroundStyle(theme.palette.textPrimary)
            }
        }
    }

    @ViewBuilder
    private func chooser(
        identity: SmbIdentity,
        path: String,
        entries: [SmbEntry]
    ) -> some View {
        Section {
            if let parent = connection.parent(of: path) {
                Button {
                    Task { await connection.enter(parent) }
                } label: {
                    Label(
                        String(localized: "smb.up", bundle: .module, locale: .storyArc),
                        systemImage: "arrow.up.left"
                    )
                }
            }
            ForEach(entries.filter(\.isDirectory)) { folder in
                Button {
                    Task { await connection.enter(folder.path) }
                } label: {
                    Label(folder.name, systemImage: "folder")
                }
            }
        } header: {
            Text(path.isEmpty ? String(localized: "smb.root", bundle: .module, locale: .storyArc) : path)
        } footer: {
            // `network-share`: the detail screen states whether the connection is
            // encrypted. Said here too, because this is the moment a reader decides
            // whether to trust it.
            Text(
                identity.isEncrypted
                    ? "smb.encrypted \(identity.dialect)"
                    : "smb.notEncrypted \(identity.dialect)",
                bundle: .module
            )
        }

        Section {
            // The other half of the same journey, and the same argument: this is the action
            // that adds the library, and a list row full of folder rows is where it must not
            // look like one more folder.
            Button {
                connection.source().map(onAdd)
                dismiss()
            } label: {
                Text("smb.useFolder", bundle: .module)
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, StoryArcSpace.xs)
            }
            .buttonStyle(.glassProminent)
            .controlSize(.large)
            .tint(theme.accent)
            .listRowBackground(Color.clear)
            .listRowInsets(EdgeInsets())
        }
    }
}
